crates/red_knot_python_semantic/src/types.rs

```diff
@@ -2318,36 +2301,37 @@ impl<'db> Type<'db> {
         }
 
         let return_ty = match self.try_call_dunder(db, "__len__", &CallArguments::none()) {
-            Ok(outcome) | Err(CallDunderError::PossiblyUnbound(outcome)) => outcome.return_type(db),
+            Ok(bindings) => bindings.return_type(db),
+            Err(CallDunderError::PossiblyUnbound(bindings)) => bindings.return_type(db),
 
             // TODO: emit a diagnostic
-            Err(err) => err.return_type(db)?,
+            Err(CallDunderError::MethodNotAvailable) => return None,
+            Err(CallDunderError::CallError(_, bindings)) => bindings.return_type(db),
         };
 
         non_negative_int_literal(db, return_ty)
     }
 
-    /// Calls `self`
+    /// Returns the call signatures of a type.
     ///
-    /// Returns `Ok` if the call with the given arguments is successful and `Err` otherwise.
-    fn try_call(
-        self,
-        db: &'db dyn Db,
-        arguments: &CallArguments<'_, 'db>,
-    ) -> Result<CallOutcome<'db>, CallError<'db>> {
+    /// Note that all types have a valid [`Signatures`], even if the type is not callable.
+    /// Moreover, "callable" can be subtle for a union type, since some union elements might be
+    /// callable and some not. A union is callable if every element type is callable — and even
+    /// then, the elements might be inconsistent, such that there's no argument list that's valid
+    /// for all elements. It's usually best to only worry about "callability" relative to a
+    /// particular argument list, via [`try_call`][Self::try_call] and
+    /// [`CallErrorKind::NotCallable`].
+    #[salsa::tracked(return_ref)]
+    fn signatures(self, db: &'db dyn Db, callable_ty: Type<'db>) -> Signatures<'db> {
```

Comment:
It feels to me like it should be possible to get rid of this `callable_ty` argument (and thus the Salsa interning) entirely, but it might require a recursive `Signature::set_callable_ty` method; not sure if that's worth it or not.